File: frontend/src/services/aiHealthService.ts
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080';

export interface AIAnalysisResult {
  id?: string;
  userId?: number;
  condition: string;
  confidence: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  description: string;
  recommendations: string[];
  imageUrl?: string;
  analyzedAt: string;
}

export interface SymptomAnalysisRequest {
  userId: number;
  symptoms: string;
}

export interface SymptomAnalysisResult {
  id?: string;
  userId?: number;
  symptoms: string;
  possibleConditions: string[];
  riskLevel: string;
  recommendations: string[];
  requiresImmediateAttention: boolean;
  analyzedAt: string;
}

export interface VoiceAnalysisResult {
  id?: string;
  userId?: number;
  transcription: string;
  response: string;
  audioUrl?: string;
  analyzedAt: string;
}

export interface AnalysisHistoryItem {
  id: string;
  userId: number;
  condition: string;
  confidence: number;
  severity: string;
  description: string;
  recommendations: string[];
  imageUrl?: string;
  analyzedAt: string;
}

class AIHealthService {
  private baseURL = `${API_BASE_URL}/api/ai-health`;

  async analyzeImage(
    image: File,
    userId: number,
    specialty: string = 'chest',
    modelType: string = 'base'
  ): Promise<AIAnalysisResult> {
    const formData = new FormData();
    formData.append('image', image);
    formData.append('userId', userId.toString());
    formData.append('specialty', specialty);
    formData.append('model_type', modelType);

    const response = await fetch(`${this.baseURL}/analyze-image`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      throw new Error(`Image analysis failed: ${response.statusText}`);
    }

    const data = await response.json();
    return this.mapAnalysisResponse(data);
  }

  async analyzeSymptoms(request: SymptomAnalysisRequest): Promise<SymptomAnalysisResult> {
    const response = await fetch(`${this.baseURL}/analyze-symptoms`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      throw new Error(`Symptom analysis failed: ${response.statusText}`);
    }

    return await response.json();
  }

  async analyzeVoice(audio: File, userId: number): Promise<VoiceAnalysisResult> {
    const formData = new FormData();
    formData.append('audio', audio);
    formData.append('userId', userId.toString());

    const response = await fetch(`${this.baseURL}/voice-analysis`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      throw new Error(`Voice analysis failed: ${response.statusText}`);
    }

    return await response.json();
  }

  async getAnalysisHistory(userId: number): Promise<AnalysisHistoryItem[]> {
    const response = await fetch(`${this.baseURL}/analysis-history/${userId}`);
    
    if (!response.ok) {
      throw new Error(`Failed to fetch analysis history: ${response.statusText}`);
    }

    const data = await response.json();
    return data.map(this.mapAnalysisResponse);
  }

  private mapAnalysisResponse(data: any): AIAnalysisResult {
    return {
      id: data.id?.toString(),
      userId: data.userId,
      condition: data.condition,
      confidence: data.confidence,
      severity: data.severity,
      description: data.description,
      recommendations: data.recommendations || [],
      imageUrl: data.imageUrl,
      analyzedAt: data.analyzedAt,
    };
  }

  // Utility method to get mock user ID (replace with real auth)
  getCurrentUserId(): number {
    // TODO: Replace with real authentication
    return 1;
  }

  // Helper method to format confidence as percentage
  formatConfidence(confidence: number): string {
    return `${Math.round(confidence)}%`;
  }

  // Helper method to get severity color
  getSeverityColor(severity: string): string {
    switch (severity) {
      case 'low': return 'text-green-600 bg-green-100';
      case 'medium': return 'text-yellow-600 bg-yellow-100';
      case 'high': return 'text-orange-600 bg-orange-100';
      case 'critical': return 'text-red-600 bg-red-100';
      default: return 'text-gray-600 bg-gray-100';
    }
  }

  // Helper method to get specialty display name
  getSpecialtyDisplayName(specialty: string): string {
    switch (specialty) {
      case 'chest': return 'Chest & Respiratory';
      case 'dermatology': return 'Dermatology';
      case 'ophthalmology': return 'Ophthalmology';
      default: return specialty;
    }
  }

  // Helper method to get model display name
  getModelDisplayName(modelType: string): string {
    switch (modelType) {
      case 'small': return 'Small (Fast)';
      case 'base': return 'Base (Balanced)';
      case 'large': return 'Large (Accurate)';
      default: return modelType;
    }
  }
}

export const aiHealthService = new AIHealthService();
export default aiHealthService;

File: backend/src/main/java/com/healthcare/service/AIHealthService.java
package com.healthcare.service;

import com.healthcare.dto.AIAnalysisResultDTO;
import com.healthcare.dto.SymptomAnalysisDTO;
import com.healthcare.dto.VoiceAnalysisDTO;
import com.healthcare.entity.AIAnalysisResult;
import com.healthcare.entity.VoiceSession;
import com.healthcare.repository.AIAnalysisResultRepository;
import com.healthcare.repository.VoiceSessionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.*;

@Service
public class AIHealthService {

    @Autowired
    private AIAnalysisResultRepository analysisRepository;

    @Autowired
    private VoiceSessionRepository voiceSessionRepository;

    @Autowired
    private VertexAIService vertexAIService;

    @Autowired
    private VAPIService vapiService;

    @Autowired
    private CloudStorageService cloudStorageService;

    public AIAnalysisResultDTO analyzeImage(MultipartFile image, Long userId, String specialty, String modelType) throws Exception {
        long startTime = System.currentTimeMillis();
        
        try {
            // Upload image to secure cloud storage
            String imageUrl;
            if (cloudStorageService.isCloudStorageAvailable()) {
                imageUrl = cloudStorageService.uploadMedicalImage(image, userId, "image-analysis");
            } else {
                imageUrl = cloudStorageService.saveFileLocally(image, "images");
            }
            
            // Analyze image with Vertex AI
            Map<String, Object> analysisResult = vertexAIService.analyzeMedicalImage(
                image.getBytes(), specialty, modelType);
            
            // Create result DTO
            AIAnalysisResultDTO result = new AIAnalysisResultDTO();
            result.setUserId(userId);
            result.setCondition((String) analysisResult.get("prediction"));
            result.setConfidence((Double) analysisResult.get("confidence"));
            result.setSeverity((String) analysisResult.get("severity"));
            result.setDescription((String) analysisResult.get("description"));
            result.setRecommendations((List<String>) analysisResult.get("recommendations"));
            result.setImageUrl(imageUrl);
            result.setAnalyzedAt(LocalDateTime.now());
            
            // Save to database
            AIAnalysisResult entity = convertToEntity(result, userId);
            entity.setAnalysisType(AIAnalysisResult.AnalysisType.IMAGE);
            entity.setSpecialty(specialty);
            entity.setModelType(modelType);
            entity.setImageUrl(imageUrl);
            entity.setProcessingTimeMs((int) (System.currentTimeMillis() - startTime));
            
            analysisRepository.save(entity);
            
            return result;
            
        } catch (Exception e) {
            // Fallback to simulation if AI service fails
            return simulateImageAnalysis(image, userId, specialty, modelType, startTime);
        }
    }

    public SymptomAnalysisDTO analyzeSymptoms(SymptomAnalysisDTO request) throws Exception {
        long startTime = System.currentTimeMillis();
        
        try {
            // Analyze symptoms with Vertex AI
            Map<String, Object> analysisResult = vertexAIService.analyzeSymptoms(
                request.getSymptoms(), request.getUserId());
            
            // Create result DTO
            SymptomAnalysisDTO result = new SymptomAnalysisDTO();
            result.setUserId(request.getUserId());
            result.setSymptoms(request.getSymptoms());
            result.setPossibleConditions((List<String>) analysisResult.get("possible_conditions"));
            result.setRiskLevel((String) analysisResult.get("risk_level"));
            result.setRecommendations((List<String>) analysisResult.get("recommendations"));
            result.setRequiresImmediateAttention((Boolean) analysisResult.get("requires_immediate_attention"));
            result.setAnalyzedAt(LocalDateTime.now());
            
            // Save to database
            AIAnalysisResult entity = new AIAnalysisResult();
            entity.setUserId(request.getUserId());
            entity.setAnalysisType(AIAnalysisResult.AnalysisType.SYMPTOM);
            entity.setSymptomsText(request.getSymptoms());
            entity.setPrediction(result.getPossibleConditions().get(0));
            entity.setConfidence(85.0); // Default confidence for symptoms
            entity.setSeverity(mapRiskLevelToSeverity(result.getRiskLevel()));
            entity.setRecommendations(result.getRecommendations());
            entity.setProcessingTimeMs((int) (System.currentTimeMillis() - startTime));
            
            analysisRepository.save(entity);
            
            return result;
            
        } catch (Exception e) {
            // Fallback to simulation if AI service fails
            return simulateSymptomAnalysis(request, startTime);
        }
    }

    public VoiceAnalysisDTO analyzeVoice(MultipartFile audio, Long userId) throws Exception {
        long startTime = System.currentTimeMillis();
        
        try {
            // Process voice with VAPI
            Map<String, Object> voiceResult = vapiService.processVoiceInput(
                audio.getBytes(), userId);
            
            // Upload audio to secure storage
            String audioUrl;
            String sessionId = (String) voiceResult.get("session_id");
            if (cloudStorageService.isCloudStorageAvailable()) {
                audioUrl = cloudStorageService.uploadAudioFile(audio, userId, sessionId);
            } else {
                audioUrl = cloudStorageService.saveFileLocally(audio, "audio");
            }
            
            // Create result DTO
            VoiceAnalysisDTO result = new VoiceAnalysisDTO();
            result.setUserId(userId);
            result.setTranscription((String) voiceResult.get("transcription"));
            result.setResponse((String) voiceResult.get("ai_response"));
            result.setAudioUrl(audioUrl);
            result.setAnalyzedAt(LocalDateTime.now());
            
            // Save voice session
            VoiceSession session = new VoiceSession();
            session.setUserId(userId);
            session.setSessionId(sessionId);
            session.setTranscript(result.getTranscription());
            session.setAiResponse(result.getResponse());
            session.setAudioUrl(audioUrl);
            session.setDurationSeconds((Integer) voiceResult.get("duration_seconds"));
            
            voiceSessionRepository.save(session);
            
            return result;
            
        } catch (Exception e) {
            // Fallback to simulation if VAPI service fails
            return simulateVoiceAnalysis(audio, userId, startTime);
        }
    }

    public List<AIAnalysisResultDTO> getAnalysisHistory(Long userId) {
        List<AIAnalysisResult> results = analysisRepository.findByUserIdOrderByCreatedAtDesc(userId);
        return results.stream()
                .map(this::convertToDTO)
                .toList();
    }

    // Fallback simulation methods (used when AI services are unavailable)
    private AIAnalysisResultDTO simulateImageAnalysis(MultipartFile image, Long userId, String specialty, String modelType, long startTime) throws IOException {
        String imageUrl = cloudStorageService.saveFileLocally(image, "images");
        
        Random random = new Random();
        double baseConfidence = switch (modelType) {
            case "small" -> 70 + random.nextDouble() * 15; // 70-85%
            case "base" -> 75 + random.nextDouble() * 20;  // 75-95%
            case "large" -> 80 + random.nextDouble() * 20; // 80-100%
            default -> 75 + random.nextDouble() * 20;
        };
        
        Map<String, String[]> specialtyConditions = Map.of(
            "chest", new String[]{"Normal Chest X-ray", "Possible Pneumonia", "Lung Nodule Detected", "Pleural Effusion"},
            "dermatology", new String[]{"Benign Skin Lesion", "Suspicious Lesion - Requires Evaluation", "Melanoma Risk", "Basal Cell Carcinoma"},
            "ophthalmology", new String[]{"Normal Retina", "Diabetic Retinopathy Signs", "Macular Degeneration", "Glaucoma Indicators"}
        );
        
        String[] conditions = specialtyConditions.getOrDefault(specialty, specialtyConditions.get("chest"));
        String condition = conditions[random.nextInt(conditions.length)];
        String severity = baseConfidence > 85 ? "low" : baseConfidence > 70 ? "medium" : "high";
        
        List<String> recommendations = new ArrayList<>();
        if (baseConfidence > 85) {
            recommendations.addAll(Arrays.asList(
                "Continue regular health checkups",
                "Maintain healthy lifestyle",
                "Monitor for any changes"
            ));
        } else {
            recommendations.addAll(Arrays.asList(
                "Consult with a " + getSpecialtyDoctor(specialty) + " immediately",
                "Follow up with additional tests",
                "Monitor symptoms closely"
            ));
        }
        
        AIAnalysisResultDTO result = new AIAnalysisResultDTO();
        result.setUserId(userId);
        result.setCondition(condition);
        result.setConfidence(baseConfidence);
        result.setSeverity(severity);
        result.setDescription("AI analysis using " + modelType + " model for " + specialty + " specialty shows " + condition.toLowerCase());
        result.setRecommendations(recommendations);
        result.setImageUrl(imageUrl);
        result.setAnalyzedAt(LocalDateTime.now());
        
        // Save to database
        AIAnalysisResult entity = convertToEntity(result, userId);
        entity.setAnalysisType(AIAnalysisResult.AnalysisType.IMAGE);
        entity.setSpecialty(specialty);
        entity.setModelType(modelType);
        entity.setImageUrl(imageUrl);
        entity.setProcessingTimeMs((int) (System.currentTimeMillis() - startTime));
        
        analysisRepository.save(entity);
        
        return result;
    }

    private SymptomAnalysisDTO simulateSymptomAnalysis(SymptomAnalysisDTO request, long startTime) {
        String symptoms = request.getSymptoms().toLowerCase();
        
        List<String> possibleConditions = new ArrayList<>();
        String riskLevel = "low";
        List<String> recommendations = new ArrayList<>();
        
        if (symptoms.contains("fever") || symptoms.contains("cough")) {
            possibleConditions.add("Upper Respiratory Infection");
            possibleConditions.add("Common Cold");
            riskLevel = "medium";
            recommendations.addAll(Arrays.asList(
                "Rest and stay hydrated",
                "Monitor temperature",
                "Consult doctor if symptoms worsen"
            ));
        } else if (symptoms.contains("chest pain") || symptoms.contains("shortness of breath")) {
            possibleConditions.add("Possible Cardiac Issue");
            riskLevel = "high";
            recommendations.addAll(Arrays.asList(
                "Seek immediate medical attention",
                "Do not ignore chest pain",
                "Call emergency services if severe"
            ));
        } else {
            possibleConditions.add("General Health Concern");
            recommendations.addAll(Arrays.asList(
                "Monitor symptoms",
                "Consult healthcare provider",
                "Maintain healthy lifestyle"
            ));
        }
        
        SymptomAnalysisDTO result = new SymptomAnalysisDTO();
        result.setUserId(request.getUserId());
        result.setSymptoms(request.getSymptoms());
        result.setPossibleConditions(possibleConditions);
        result.setRiskLevel(riskLevel);
        result.setRecommendations(recommendations);
        result.setRequiresImmediateAttention("high".equals(riskLevel));
        result.setAnalyzedAt(LocalDateTime.now());
        
        // Save to database
        AIAnalysisResult entity = new AIAnalysisResult();
        entity.setUserId(request.getUserId());
        entity.setAnalysisType(AIAnalysisResult.AnalysisType.SYMPTOM);
        entity.setSymptomsText(request.getSymptoms());
        entity.setPrediction(result.getPossibleConditions().get(0));
        entity.setConfidence(85.0);
        entity.setSeverity(mapRiskLevelToSeverity(result.getRiskLevel()));
        entity.setRecommendations(result.getRecommendations());
        entity.setProcessingTimeMs((int) (System.currentTimeMillis() - startTime));
        
        analysisRepository.save(entity);
        
        return result;
    }

    private VoiceAnalysisDTO simulateVoiceAnalysis(MultipartFile audio, Long userId, long startTime) throws IOException {
        String audioUrl = cloudStorageService.saveFileLocally(audio, "audio");
        
        String[] sampleTranscripts = {
            "I have been feeling tired and have a headache for the past two days",
            "My throat is sore and I have been coughing",
            "I'm experiencing chest pain and shortness of breath"
        };
        
        String[] sampleResponses = {
            "Based on your symptoms, you may be experiencing fatigue and tension headaches. I recommend rest, hydration, and monitoring your symptoms.",
            "Your symptoms suggest a possible upper respiratory infection. Please rest, stay hydrated, and consider seeing a doctor if symptoms persist.",
            "Chest pain and shortness of breath require immediate medical attention. Please seek emergency care right away."
        };
        
        Random random = new Random();
        int index = random.nextInt(sampleTranscripts.length);
        
        VoiceAnalysisDTO result = new VoiceAnalysisDTO();
        result.setUserId(userId);
        result.setTranscription(sampleTranscripts[index]);
        result.setResponse(sampleResponses[index]);
        result.setAudioUrl(audioUrl);
        result.setAnalyzedAt(LocalDateTime.now());
        
        // Save voice session
        VoiceSession session = new VoiceSession();
        session.setUserId(userId);
        session.setSessionId(UUID.randomUUID().toString());
        session.setTranscript(result.getTranscription());
        session.setAiResponse(result.getResponse());
        session.setAudioUrl(audioUrl);
        session.setDurationSeconds((int) (System.currentTimeMillis() - startTime) / 1000);
        
        voiceSessionRepository.save(session);
        
        return result;
    }

    private String getSpecialtyDoctor(String specialty) {
        return switch (specialty) {
            case "chest" -> "pulmonologist";
            case "dermatology" -> "dermatologist";
            case "ophthalmology" -> "ophthalmologist";
            default -> "specialist";
        };
    }

    private AIAnalysisResult convertToEntity(AIAnalysisResultDTO dto, Long userId) {
        AIAnalysisResult entity = new AIAnalysisResult();
        entity.setUserId(userId);
        entity.setPrediction(dto.getCondition());
        entity.setConfidence(dto.getConfidence());
        entity.setSeverity(AIAnalysisResult.Severity.valueOf(dto.getSeverity().toUpperCase()));
        entity.setDescription(dto.getDescription());
        entity.setRecommendations(dto.getRecommendations());
        return entity;
    }

    private AIAnalysisResultDTO convertToDTO(AIAnalysisResult entity) {
        AIAnalysisResultDTO dto = new AIAnalysisResultDTO();
        dto.setId(entity.getId());
        dto.setUserId(entity.getUserId());
        dto.setCondition(entity.getPrediction());
        dto.setConfidence(entity.getConfidence());
        dto.setSeverity(entity.getSeverity().name().toLowerCase());
        dto.setDescription(entity.getDescription());
        dto.setRecommendations(entity.getRecommendations());
        dto.setImageUrl(entity.getImageUrl());
        dto.setAnalyzedAt(entity.getCreatedAt());
        return dto;
    }

    private AIAnalysisResult.Severity mapRiskLevelToSeverity(String riskLevel) {
        return switch (riskLevel.toLowerCase()) {
            case "low" -> AIAnalysisResult.Severity.LOW;
            case "medium" -> AIAnalysisResult.Severity.MEDIUM;
            case "high" -> AIAnalysisResult.Severity.HIGH;
            case "critical" -> AIAnalysisResult.Severity.CRITICAL;
            default -> AIAnalysisResult.Severity.LOW;
        };
    }
}